src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions
+{
+    using RegexNodeOrToken = EmbeddedSyntaxNodeOrToken<RegexKind, RegexNode>;
+    using RegexToken = EmbeddedSyntaxToken<RegexKind>;
+
+    internal sealed class RegexCompilationUnit : RegexNode
+    {
+        public RegexCompilationUnit(RegexExpressionNode expression, RegexToken endOfFileToken)
+            : base(RegexKind.CompilationUnit)
+        {
+            Debug.Assert(expression != null);
+            Debug.Assert(endOfFileToken.Kind == RegexKind.EndOfFile);
+            Expression = expression;
+            EndOfFileToken = endOfFileToken;
+        }
+
+        public RegexExpressionNode Expression { get; }
+        public RegexToken EndOfFileToken { get; }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return Expression;
+            case 1: return EndOfFileToken;
+            }
+
+            throw new InvalidOperationException();
```

Comment:
Maybe out of range exception instead?

Reply:
#won'tfix.  I think InvalidOperation is appropriate for this sort of thing.

Reply:
So, to me, this is hte function you call after already calling ChildCount.  Indeed, it's only really for use by the node Enumerator code.  So, the only way you wouldn't get one of the normal cases is if somehow ChildCount and ChildAt were out of sync indicating that our invariants themselves are out of whack, and less that there's a bug in the caller.

That said, i don't feel strongly.  and will change if you both prefer that :)